internal/librarian/update_image_test.go

```diff
@@ -0,0 +1,762 @@
+					return err
+				}
+				return nil
+			},
+		},
+		{
+			name: "valid config with local repo",
+			cfg: &config.Config{
+				API:         "some/api",
+				APISource:   newTestGitRepo(t).GetDir(),
+				Branch:      "test-branch",
+				Repo:        newTestGitRepo(t).GetDir(),
+				WorkRoot:    t.TempDir(),
+				Image:       "gcr.io/test/test-image",
+				CommandName: updateImageCmdName,
+			},
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			// custom setup
+			if test.setupFunc != nil {
+				if err := test.setupFunc(test.cfg); err != nil {
+					t.Fatalf("error in setup %v", err)
+				}
+			}
+
+			r, err := newUpdateImageRunner(test.cfg)
+			if test.wantErr {
+				if err == nil {
+					t.Fatalf("newUpdateImageRunner() error = %v, wantErr %v", err, test.wantErr)
+				}
+
+				if !strings.Contains(err.Error(), test.wantErrMsg) {
+					t.Fatalf("want error message: %s, got: %s", test.wantErrMsg, err.Error())
+				}
+
+				return
+			}
+
+			if err != nil {
+				t.Fatalf("newUpdateImageRunner() got error: %v", err)
+			}
+
+			if r.branch == "" {
+				t.Errorf("newUpdateImageRunner() branch is not set")
+			}
+
+			if r.ghClient == nil {
+				t.Errorf("newUpdateImageRunner() ghClient is nil")
+			}
+			if r.containerClient == nil {
+				t.Errorf("newUpdateImageRunner() containerClient is nil")
+			}
+			if r.repo == nil {
+				t.Errorf("newUpdateImageRunner() repo is nil")
+			}
+			if r.sourceRepo == nil {
+				t.Errorf("newUpdateImageRunner() sourceRepo is nil")
+			}
+		})
+	}
+}
+
+func TestUpdateImageRunnerRun(t *testing.T) {
+	t.Parallel()
+	for _, test := range []struct {
+		name                       string
+		imagesClient               *mockImagesClient
+		containerClient            *mockContainerClient
+		ghClient                   *mockGitHubClient
+		state                      *config.LibrarianState
+		image                      string
+		build                      bool
+		commit                     bool
+		push                       bool
+		wantErr                    bool
+		wantErrMsg                 string
+		wantFindLatestCalls        int
+		wantGenerateCalls          int
+		wantBuildCalls             int
+		wantCheckoutCalls          int
+		wantCreatePullRequestCalls int
+		wantCreateIssueCalls       int
+		wantCommitMsg              string
+		checkoutError              error
+	}{
+		{
+			name:  "specific image",
+			image: "some-image",
+			state: &config.LibrarianState{
+				Image: "gcr.io/test/image:v1.2.3",
+				Libraries: []*config.LibraryState{
+					{
+						ID:   "lib1",
+						APIs: []*config.API{{Path: "some/api1"}},
+						SourceRoots: []string{
+							"src/a",
+						},
+						LastGeneratedCommit: "abcd1234",
+					},
+				},
+			},
+			containerClient:     &mockContainerClient{},
+			imagesClient:        &mockImagesClient{},
+			ghClient:            &mockGitHubClient{},
+			wantFindLatestCalls: 0,
```

Comment:
nit: prefer not including fields if they are the default values, so anywhere there is a 0 or maybe an empty string we can omit that line.

Reply:
I made this one explicit because we're asserting that we don't make any calls to the API to find the latest image if one is explicitly provided. I think it makes this test case easier to understand.